### PR TITLE
Faucet should be limited

### LINKED ISSUE #15: Faucet should be limited
## User Story
As a faucet operator
I want to limit the amount of funds an account can request
Because I want to be more fair in funds distribution

## Scenarios
**Scenario 1 (in line with the current behavior)**
Given that an account is **not** funded above the limit
When a fill request for that account is made
Then the account is funded the configured amount

**Scenario 2**
Given that an account **is** funded above the limit
When a fill request for that account is made
Then the account is **not** funded
And the server returns an understandable error to the client

## Info

* We understand that this system is not safe from abuse. An attacker could create any number of accounts and fund them.
* Another approach to funding limitation would be a rate limit. However, a rate limit requires a much more sophisticated implementation. It is currently not worth the effort, but may be added at a later stage.
* The maximum funding should be set in the configuration of the faucet.
* The faucet should check the balance (on-chain with the node) before funding an account.

### PR DESCRIPTION
**Pull request by 0xsarvesh**
fixes #15 

This PR adds rate limiting by checking beneficiary balances. 

if the beneficiary already has enough balance defined in the config then faucet will not fund it.

### CODE REVIEW

**Review by schemar — APPROVED** [review]
Very nice change 🦖
Someone else should review as well, please.

I like the new Exception and the linting fixed you did in addition 👍

Some comments which are not critical:

* Maybe it makes sense to start moving code from the two faucets into a shared parent class.
* Maybe you could extend the existing faucets into „LimitedErc20Faucet“ and coin faucet. That way you would not need to change the existing faucets. But I think it’s better the way you did it here 🤔 faucets should probably always be limited.

File: src/Server.ts [hunk 1]
```diff
@@ -166,10 +174,18 @@ export default class Server {
...
+
+      if (!(error instanceof EnoughFundException)) {
+        return Server.returnError(
+          response,
+          new ServerError('Server error. Could not fill address.', 500),
+        );
+      } else {
+        return Server.returnError(
+          response,
+          new ServerError(`${error.toString()}`, 422),
```
Comment (deepesh-kn) [inline src/Server.ts]: We should have the status code as a constants or Enums

File: src/Server.ts [hunk 2]
```diff
@@ -123,6 +124,13 @@ export default class Server {
     response: http.ServerResponse,
   ): Promise<void> {
     let body: Uint8Array[];
+
+    /* Allow cross origin request, this is needed to build an UI on top of faucet */
```
Comment (deepesh-kn) [inline src/Server.ts]: We should think about this if we want to allow the cross-origin request.
Maybe we can deploy the UI at a particular domain and just allow access for that domain.

**Review by deepesh-kn — COMMENTED** [review]
Nicely done. 💯 

The only concern here is about enabling CORS. We can disable the CORS and should come back to it once we deploy the UI. What do you think?

I like @schemar's comments 👍
For now, we can go with these changes.

File: src/Server.ts (hunk 2 above)
Comment (schemar) [reply src/Server.ts]: Shouldn't the headers only be sent in response to an `OPTIONS` request? See [Preflight requests in CORS on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS#Preflighted_requests_in_CORS)

Comment (0xsarvesh) [reply src/Server.ts]: @schemar I am always confused on this topic. Please check the second para on the link, it says all the methods.  Though I am not sure about this. 

> The server responds with Access-Control-Allow-Methods and says that POST, GET, and OPTIONS are viable methods to query the resource in question. This header is similar to the Allow response header, but used strictly within the context of CORS.

**Comment by 0xsarvesh** [comment]
> Nicely done. 💯
> 
> The only concern here is about enabling CORS. We can disable the CORS and should come back to it once we deploy the UI. What do you think?

Enabling CORS will accept requests from the browser also which is fine because the faucet is anyways open(Anyone can make curl request). Faucet is limited with different logic(Based on balance).  
> 
> I like @schemar's comments 👍
> For now, we can go with these changes.

**Comment by 0xsarvesh** [comment]
I have updated faucet to read environment variables for `CORS` settings. It's not hardcoded in the code anymore.

**Review by deepesh-kn — APPROVED** [review]
LGTM :shipit:

File: src/Server.ts (hunk 2 above)
Comment (schemar) [reply src/Server.ts]: You make an `OPTIONS` **request**. In the **response** to *that* request, the server tells the client which methods are allowed (for *future* requests), e.g. `OPTIONS` and `GET`.

**Pull request merged**